synapse/lib/stormlib/imap.py

```diff
@@ -0,0 +1,376 @@
+                     'desc': 'The IMAP server port.'},
+                    {'type': 'int', 'name': 'timeout', 'default': 30,
+                     'desc': 'The time to wait for all commands on the server to execute.'},
+                    {'type': 'bool', 'name': 'ssl', 'default': True,
+                     'desc': 'Use SSL to connect to the IMAP server.'},
+                ),
+                'returns': {
+                    'type': 'storm:imap:server',
+                    'desc': 'A new ``storm:imap:server`` instance.'
+                },
+            },
+        },
+    )
+    _storm_lib_path = ('inet', 'imap', )
+
+    def getObjLocals(self):
+        return {
+            'server': self.server,
+        }
+
+    async def server(self, host, port=993, timeout=30, ssl=True):
+        ssl = await s_stormtypes.tobool(ssl)
+        host = await s_stormtypes.tostr(host)
+        port = await s_stormtypes.toint(port)
+        timeout = await s_stormtypes.toint(timeout, noneok=True)
+
+        if ssl:
+            imap_cli = aioimaplib.IMAP4_SSL(host=host, port=port, timeout=timeout)
+        else:
+            imap_cli = aioimaplib.IMAP4(host=host, port=port, timeout=timeout)
+
+        async def fini():
+            # call protocol.logout() so fini() doesn't hang
+            await asyncio.wait_for(imap_cli.protocol.logout(), 5)
+
+        self.runt.snap.onfini(fini)
+
+        try:
+            await imap_cli.wait_hello_from_server()
+        except asyncio.TimeoutError:
+            raise s_exc.TimeOut(mesg='Timed out waiting for IMAP server hello.') from None
+
+        return ImapServer(self.runt, imap_cli)
+
+@s_stormtypes.registry.registerType
+class ImapServer(s_stormtypes.StormType):
+    '''
+    An IMAP server for retrieving email messages.
+    '''
+    _storm_locals = (
+        {
+            'name': 'list',
+            'desc': '''
+            List mailbox names.
+
+            By default this method uses a reference_name and pattern to return
+            all mailboxes from the root.
+            ''',
+            'type': {
+                'type': 'function', '_funcname': 'list',
+                'args': (
+                    {'type': 'str', 'name': 'reference_name', 'default': '""',
+                     'desc': 'The mailbox reference name.'},
+                    {'type': 'str', 'name': 'pattern', 'default': '*',
+                     'desc': 'The pattern to filter by.'},
+                ),
+                'returns': {
+                    'type': 'list',
+                    'desc': 'An ($ok, $valu) tuple where $valu is a list of names if $ok=True.'
+                },
+            },
+        },
+        {
+            'name': 'fetch',
+            'desc': '''
+            Fetch a message by UID in RFC822 format.
+
+            The message is saved to the Axon, and a file:bytes node is returned.
+
+            Examples:
+                Fetch a message, save to the Axon, and yield ``file:bytes`` node::
```

Comment:
Great call on saving to the axon and avoiding having the bytes in the storm runtime